### PR TITLE
meta: add appstream metadata extractor

### PR DESCRIPTION
**Pull request by cmatsuoka**
Extract application metadata from appstream files.

Co-authored-by: Kyle Fazzari <kyle@canonical.com>
Co-authored-by: Leo Arias <leo.arias@canonical.com>
Co-authored-by: Sergio Schvezov <sergio.schvezov@canonical.com>
Signed-off-by: Claudio Matsuoka <claudio.matsuoka@canonical.com>

- [x] Have you followed the [guidelines for contributing](https://github.com/snapcore/snapcraft/blob/master/CONTRIBUTING.md)?
- [x] Have you signed the [CLA](http://www.ubuntu.com/legal/contributors/)?
- [ ] Have you successfully run `./runtests.sh static`?
- [ ] Have you successfully run `./runtests.sh tests/unit`?

-----
CRAFT-1033

### CODE REVIEW

File: snapcraft/errors.py [hunk 1]
```diff
@@ -42,5 +42,12 @@ class ExtensionError(SnapcraftError):
     """Error during parts processing."""
 
 
+class MetadataExtractionError(SnapcraftError):
+    """Attempt to extract metadata from file was unsuccessful."""
+
+    def __init__(self, filename: str) -> None:
+        super().__init__(f"Error extracting metadata from {filename}")
```
Comment (sergiusens) [inline snapcraft/errors.py:49]:
```suggestion
        super().__init__(f"Error extracting metadata from {filename!r}")
```

File: snapcraft/meta/appstream.py [hunk 2]
```diff
@@ -0,0 +1,276 @@
...
+    dom = _get_dom(path)
+    transform = _get_xslt()
+    return transform(dom)
+
+
+def _get_dom(path: str) -> lxml.etree.ElementTree:
+    try:
+        return lxml.etree.parse(path)
+    except lxml.etree.ParseError as err:
+        raise errors.MetadataExtractionError(path) from err
```
Comment (sergiusens) [inline snapcraft/meta/appstream.py:142]: is there a test for this condition?

**Review by sergiusens — APPROVED** [review]
Looks good, a couple of side comments

File: snapcraft/meta/appstream.py (hunk 2 above)
Comment (cmatsuoka) [reply snapcraft/meta/appstream.py:142]: Added an extraction error test.

File: snapcraft/errors.py (hunk 1 above)
Comment (cmatsuoka) [reply snapcraft/errors.py:49]: Amended as suggested.

**Pull request merged**